### PR TITLE
relay-example: Add location mutation

### PR DESCRIPTION
**Pull request by tbergq**
- Add RootMutation to schema
- Add Error interace type
- Create teh addLocationMutation in graphql

### CODE REVIEW

File: src/example-relay/src/graphql/locations/types/input/AddLocationInput.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+});
+
+export default new GraphQLInputObjectType({
+  name: 'AddLocationInput',
+  fields: {
+    locationId: {
+      type: GraphQLNonNull(GraphQLString),
+      description: '3-letter IATA code of airport or internal city code.',
+    },
+    name: {
```
Comment (michalsanger) [inline src/example-relay/src/graphql/locations/types/input/AddLocationInput.js]: Why not make `name` and `type` required (non null)? IMHO it makes sense to require all those fields to create new location. And you don't have to worry about input validation in the mutation resolver.

File: src/example-relay/src/graphql/locations/types/output/AddLocationError.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import ErrorInterface from '../../../types/interface/Error';
+
+export default new GraphQLObjectType({
+  name: 'AddLocationError',
+  description: 'Response type indicating an error',
+  interfaces: [ErrorInterface],
+  fields: {
+    message: {
+      type: GraphQLString,
+      resolve: () => 'You must proivde a name',
```
Comment (michalsanger) [inline src/example-relay/src/graphql/locations/types/output/AddLocationError.js]:
```suggestion
      resolve: () => 'You must provide a name',
```

File: src/example-relay/src/graphql/locations/types/output/AddLocationError.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+// @flow strict
+
+import { GraphQLObjectType, GraphQLString } from 'graphql';
+
+import ErrorInterface from '../../../types/interface/Error';
+
+export default new GraphQLObjectType({
+  name: 'AddLocationError',
```
Comment (michalsanger) [inline src/example-relay/src/graphql/locations/types/output/AddLocationError.js]: Maybe better to be more generic and call it just `InputError`?

File: src/example-relay/src/graphql/locations/types/output/AddLocationOrError.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  constructor(location: LocationInput) {
+    this.location = {
+      ...location,
+      id: location.locationId,
+    };
+  }
+}
+
+export default new GraphQLUnionType({
+  name: 'AddLocationOrError',
```
Comment (michalsanger) [inline src/example-relay/src/graphql/locations/types/output/AddLocationOrError.js]:
```suggestion
  name: 'AddLocationResult',
```

File: src/example-relay/src/graphql/locations/types/output/AddLocationMutationResponse.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+// @flow strict-local
+
+import { GraphQLObjectType } from 'graphql';
+
+import { LocationEdge } from './LocationConnection';
+
+export default new GraphQLObjectType({
+  name: 'AddLocationMutationResponse',
```
Comment (michalsanger) [inline src/example-relay/src/graphql/locations/types/output/AddLocationMutationResponse.js]:
```suggestion
  name: 'AddLocationResponse',
```

File: src/example-relay/src/graphql/locations/types/output/AddLocationMutationResponse.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  description: 'The field containing the newly create location',
```
Comment (michalsanger) [inline src/example-relay/src/graphql/locations/types/output/AddLocationMutationResponse.js]:
```suggestion
  description: 'The newly created location edge that can be automatically added into list by Relay',
```

File: src/example-relay/src/graphql/types/interface/Error.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
+// @flow strict
```
Comment (michalsanger) [inline src/example-relay/src/graphql/types/interface/Error.js]: Is the `interface` folder needed? Maybe put it one level up until it's really needed.

File: src/example-relay/src/graphql/locations/mutations/AddLocation.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  ...
+};
+export default {
+  type: AddLocationOrError,
+  description: 'Add a new location',
+  args: {
+    location: { type: GraphQLNonNull(AddLocationInput) },
+  },
+  resolve: (_: mixed, { location }: Args) => {
+    if (location.name === '' || location.name == null) {
```
Comment (michalsanger) [inline src/example-relay/src/graphql/locations/mutations/AddLocation.js]: No need for this if you make those param non null. Maybe duplicity check would be good here.

File: src/example-relay/src/graphql/locations/types/input/AddLocationInput.js (hunk 1 above)
Comment (tbergq) [reply src/example-relay/src/graphql/locations/types/input/AddLocationInput.js]: Yeah, I wanted to throw in here some examples of union return type, and some error handling. I was thinking about adding this later. But I guess I can make these non null, and still have some cases for showing of error handling.

File: src/example-relay/src/graphql/locations/types/output/AddLocationError.js (hunk 3 above)
Comment (tbergq) [reply src/example-relay/src/graphql/locations/types/output/AddLocationError.js]: Yes, but my idea was to add some more specific fields later, something to tell that IATA code is incorrect or something. We have the `Error` interface which is the more generic.

File: src/example-relay/src/graphql/locations/types/output/AddLocationOrError.js (hunk 4 above)
Comment (tbergq) [reply src/example-relay/src/graphql/locations/types/output/AddLocationOrError.js]: I was trying to follow the naming convention suggested from [this artsy](https://artsy.github.io/blog/2018/10/19/where-art-thou-my-error/) article.

**Comment by tbergq** [comment]
@michalsanger please check again

File: src/example-relay/src/graphql/locations/mutations/AddLocation.js (hunk 8 above)
Comment (tbergq) [reply src/example-relay/src/graphql/locations/mutations/AddLocation.js]: We would still need to check for empty string. If you by duplicity mean that it is already not in the location.json, I think it is a little outside of the scope of this example.

File: src/example-relay/src/graphql/locations/types/output/AddLocationOrError.js (hunk 4 above)
Comment (michalsanger) [reply src/example-relay/src/graphql/locations/types/output/AddLocationOrError.js]: OK, why not. In PD we came up with "Result" postfix so I suggested it here.

File: src/example-relay/src/graphql/locations/mutations/AddLocation.js (hunk 8 above)
Comment (michalsanger) [reply src/example-relay/src/graphql/locations/mutations/AddLocation.js]: I'd agree it's out of scope of this PR but I think it's OK for the example to have example of a backend validation.

Comment (tbergq) [reply src/example-relay/src/graphql/locations/mutations/AddLocation.js]: Sure, why not let’s add it later 👍

File: src/example-relay/schema.graphql [hunk 9]
```diff
@@ -1,10 +1,34 @@
...
+input AddLocationInput {
+  """3-letter IATA code of airport or internal city code."""
+  locationId: String!
+  name: String!
+
+  """Airport, city or country."""
+  type: LocationType!
+}
+
+union AddLocationOrError = AddLocationError | AddLocationResponse
```
Comment (mrtnzlml) [inline src/example-relay/schema.graphql]: 👍 That's ~how we do it in Pipedrive (apart from the naming convention).

**Pull request merged**